brain/detail/circuit.h

```diff
@@ -222,12 +189,335 @@ class Circuit::Impl
 
     URI getMorphologyURI(const std::string& name) const
     {
-        URI uri(_morphologySource);
+        URI uri(getMorphologySource());
         uri.setPath(uri.getPath() + "/" + name + ".h5");
         return uri;
     }
 
-    const brion::SynapseSummary& getSynapseSummary() const
+    virtual URI getMorphologySource() const = 0;
+
+    virtual const brion::URI& getCircuitSource() const = 0;
+    virtual void saveMorphologyToCache(
+        const std::string& uri, const std::string& hash,
+        neuron::MorphologyPtr morphology) const = 0;
+
+    virtual CachedMorphologies loadMorphologiesFromCache(
+        const std::set<std::string>& hashes) const = 0;
+    virtual void saveSynapsePositionsToCache(
+        const uint32_t gid, const std::string& hash,
+        const brion::SynapseMatrix& value) const = 0;
+    virtual CachedSynapses loadSynapsePositionsFromCache(
+        const Strings& keys) const = 0;
+    virtual void _findSynapsePositionsColumns() const = 0;
+    virtual const brion::SynapseSummary& getSynapseSummary() const = 0;
+    virtual const brion::Synapse& getSynapseAttributes(
+        const bool afferent) const = 0;
+    virtual const brion::Synapse& getAfferentProjectionAttributes(
+        const std::string& name) const = 0;
+    virtual const brion::Synapse* getSynapseExtra() const = 0;
+    virtual const brion::Synapse& getSynapsePositions(
+        const bool afferent) const = 0;
+    virtual const URI& getSynapseSource() const = 0;
+};
+
+class SonataCircuit : public Circuit::Impl
+{
+public:
+    explicit SonataCircuit(const URI& uri)
+        : config(uri)
+        , basePath(boost::filesystem::path(uri.getPath()).parent_path())
+    {
+        const auto& subnetworks = config.getNodes();
+        const auto& node = subnetworks.front();
+        brion::Nodes nodes(URI(basePath.string() + "/" + node.elements));
+
+        if (subnetworks.size() > 1)
+        {
+            LBWARN << "More than one subnetwork found, ignored." << std::endl;
+        }
+
+        const auto populations = nodes.getPopulationNames();
+        if (populations.size() > 1)
+        {
+            LBWARN << "More than one population found, ignored." << std::endl;
+        }
+
+        const auto population = populations.front();
+        const auto nodeGroupIDs = nodes.getNodeGroupIDs(population);
+        const auto nodeIDs = nodes.getNodeIDs(population);
+        const auto nodeGroupIndices = nodes.getNodeGroupIndices(population);
+        const size_t numNodes = nodeIDs.size();
+
+        if (nodeGroupIDs.size() > 1)
+        {
+            LBWARN << "More than one group ID found, ignored." << std::endl;
+        }
```

Comment:
This dataset has as many elements as cells. You're always going to give a warning unless your circuit has 1 neuron.
What you must check is that all IDs are 0 and throw if there's any other value.